api/server/server.go

```diff
@@ -392,7 +392,14 @@ func WithAgentFromEnv() ServerOption {
 			if err != nil {
 				return err
 			}
-			s.agent = agent.NewSyncOnly(agent.NewCachedDataAccess(ds))
+			grpcAddr := fmt.Sprintf(":%d", s.grpcListenPort)
+			delegatedAgent := agent.NewSyncOnly(agent.NewCachedDataAccess(ds))
+			_, cancel := context.WithCancel(ctx)
```

Comment:
my suggestion here might not be an elegant way of handling this, but perhaps we could save this 'ctx' in Server object, then handle/wait for it in startGears(), eg.

```
tmp, cancel := context.WithCancel(context.Background())
s.cancelCtx = tmp
```

otherwise failure of Start() in pure runner is missed...

Reply:
Good point, I'll see what I can do about the pure runner cancellation and the context.

Reply:
I'm now thinking that if we allow extensions or implementations to run servers alongside the main one then there could be more things that would generate a cancellation context, for example if we have extensions that implement their own http server (other metrics extensions maybe?).

What if we store an array of `cancelCtx`s to which extensions can append their own, and then use `reflect.Select` to dynamically wait for any of them to be `Done()`? Do you think it's a good solution or is it too convoluted?